qiskit/transpiler/passes/optimization/optimize_annotated.py

```diff
@@ -148,17 +167,220 @@ def _canonicalize(self, dag) -> Tuple[DAGCircuit, bool]:
+                    continue
+
+                # fn or bn could be already collected when considering other qubits
+                if fn in processed_nodes or bn in processed_nodes:
+                    continue
+
+                # it is possible that the same node is both at the front and at the back,
+                # it should not be collected
+                if fn == bn:
+                    continue
+
+                # have been checked before
+                if (fn, bn) in checked_node_pairs:
+                    continue
+
+                # fast check based on the arguments
+                if fn.qargs != bn.qargs or fn.cargs != bn.cargs:
+                    continue
+
+                # in the future we want to include a more precise check whether a pair
+                # of nodes are inverse
+                if fn.op == bn.op.inverse():
+                    # update q_to_f, q_to_b maps
+                    for q in fn.qargs:
+                        q_to_f[q] = None
+                    for q in bn.qargs:
+                        q_to_b[q] = None
+
+                    # see which other nodes become at the front and update information
+                    for node in dag.op_successors(fn):
+                        if node not in processed_nodes:
+                            in_degree[node] -= 1
+                            if in_degree[node] == 0:
+                                for q in node.qargs:
+                                    q_to_f[q] = node
+                                    unchecked_qubits.add(q)
+
+                    # see which other nodes become at the back and update information
+                    for node in dag.op_predecessors(bn):
+                        if node not in processed_nodes:
+                            out_degree[node] -= 1
+                            if out_degree[node] == 0:
+                                for q in node.qargs:
+                                    q_to_b[q] = node
+                                    unchecked_qubits.add(q)
+
+                    # collect and mark as processed
+                    front_block.append(fn)
+                    back_block.append(bn)
+                    processed_nodes.add(fn)
+                    processed_nodes.add(bn)
+
+                else:
+                    checked_node_pairs.add((fn, bn))
+
+        # if nothing is found, return None
+        if len(front_block) == 0:
+            return None
+
+        back_block = back_block[::-1]
+
+        # create the output DAGs
+        front_circuit = dag.copy_empty_like()
+        middle_circuit = dag.copy_empty_like()
+        back_circuit = dag.copy_empty_like()
+        front_circuit.global_phase = 0
+        back_circuit.global_phase = 0
+
+        for node in front_block:
+            front_circuit.apply_operation_back(node.op, node.qargs, node.cargs)
+
+        for node in back_block:
+            back_circuit.apply_operation_back(node.op, node.qargs, node.cargs)
```

Comment:
Strictly it's a bit more efficient to do `for node in reversed(back_block)` than `back_block = back_block[::1]; for node in back_block`, but that said - we could also just use `DAGCircuit.apply_operation_front`.

(Fwiw, calling `reversed` on a Python list to make an iterator will produce an iterator that's just as fast as the regular forwards one.)

Reply:
thanks, switching to `DAGCircuit.apply_operation_front`